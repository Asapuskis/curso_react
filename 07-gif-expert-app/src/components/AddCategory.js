import React, { useState } from 'react'
import PropTypes from 'prop-types';

export const AddCategory = ( {setCategories} ) => {

    // Creamos un useState para almacenar el valor de la caja de texto y la iniciamos a Kps
    const [inputValue, setInputValue] = useState('Kps');

    // Manejar los cambios de la caja de texto mediante useState.
    const handleInputChange = (e) => {
        setInputValue(e.target);
        console.log('handleInputChange llamado');
    }

    // Manejar el submit del formulario.
    const handleSubmit = (e) => {
        // Mediante preventDefault, evitamos que se refresque toda la página y perder los estados.
        e.preventDefault();

        console.log('handleSubmit', inputValue);

        if (inputValue.trim().length > 2) {
            setCategories( cats => [ inputValue, ...cats,]);
            setInputValue('');
            console.log('Cambios hechos');
        }
    }

    // Devolvemos un formulario para añadir la categoría a la lista.
    return (
        <form onSubmit={handleSubmit}>
            <h2>AddCategory</h2>
            <p>{inputValue}</p>
            <input
                type='text'
                value={inputValue}
                onChange={ handleInputChange }
            />
        </form>
    )
}

// Añadimos el propType requerido para la función setCategories
AddCategory.propTypes = {
    setCategories: PropTypes.func.isRequired
}
