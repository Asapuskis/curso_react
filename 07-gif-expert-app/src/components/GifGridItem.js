import React from 'react'
import PropTypes from 'prop-types';

// Acordarse de desestructurar image ya que por defecto viene como si fuese una prop, entonces o ponemos prop.image o {image}
export const GifGridItem = ( {image} ) => {
  console.log(`title: ${image.title}`)
  console.log(`url: ${image.url}`)
  return (
    <div className='card animate__animated animate__fadeIn '>
        <p>{image.title}</p>
        <img src={image.url} alt={image.title} />
    </div>
  )
}

GifGridItem.propTypes = {
  image: PropTypes.object.isRequired
}
