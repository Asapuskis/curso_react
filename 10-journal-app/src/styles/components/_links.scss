.link {
    text-decoration: none;
    color: $dark-grey;
    margin-bottom: 10px;
}

.link:hover {
    text-decoration: underline;
}