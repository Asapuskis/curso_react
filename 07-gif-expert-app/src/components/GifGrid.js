import React from 'react'
import PropTypes from 'prop-types';
import { useFetchGifs } from '../hooks/useFetchGifs';
import { GifGridItem } from './GifGridItem';

export const GifGrid = ( {category}) => {


    /////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

    // Código para mostrar los gifs utilizando los hooks useState y useEffect.

    // const [images, setImages] = useState([]);

    // // Hook useEffect. Este hook sirve para ejecutar cierto código de manera condicional.
    // // De esta manera evitaremos llamar al código múltiples veces cuando no es necesario recargar los componentes o el navegador.
    // // useEffect tiene dos argumentos:
    // // 1. El código que se va a ejecutar, en este caso getGifs()
    // // 2. Una lista de dependencias que se tienen que cumplir para que se ejecute este código. Si ponemos vacío [] se ejecutará una única vez.
    // // Si colocamos category, se llamará a este hook cuando la variable category se modifique.
    // useEffect( () => {
    //         getGifs(category)
    //             .then( imgs => setImages( imgs ));
    //     },
    //     [ category ]
    // );

    /////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

    /////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

    // Código para mostrar los gifs con un custom Hook creado expresamente para esto:
    // Lo ideal es utilizar custom hooks que llaman a hooks ya existentes y colocarlos aparte en el código, se recomienda crear carpeta hooks

    const {data, loading} = useFetchGifs(category);
    console.log(data);
    console.log(loading);

    /////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

    return (
        <>
            <h3 className='animate__animated animate__fadeIn'>{category}</h3>
            {/* Mediante un operador terniario, mostramos loading mientras carga los gifs y nada cuando los haya cargado */}
            { loading ? <p className='animate__animated animate__flash'>loading</p> : null}
            <div className='card-grid animate__animated animate__fadeIn'>
                {
                    data.map( img => (
                        <GifGridItem
                            image={img}
                            key={img.id}
                        />
                    ))
                }
            </div>
        </>
    )
}

GifGrid.propTypes = {
    category: PropTypes.string.isRequired
}