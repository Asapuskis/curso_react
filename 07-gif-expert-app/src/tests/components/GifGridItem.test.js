import { shallow } from 'enzyme';
import { GifGridItem } from "../../components/GifGridItem";

describe('Pruebas en <GifGridItem />', () => {

  const image = ({
    title: 'hola',
    url: 'https://wwhola'
  });

  // Utilizaremos shallow para garantizar recoger lo que se rendereiza.
  const wrapper = shallow( <GifGridItem image={image} /> );

  test('Debe mostrar <GifGridItem /> correctamente', () => {
      // toMatchSnapshot genera un archivo .snap. Estos archivos son fotografías de como queda el componente.
      // Previamente hay que instalar enzyme-to-json: npm install --save-dev enzyme-to-json
      expect( wrapper ).toMatchSnapshot();
  });

  test('Debe tener un parrafo con el title', () => {
      const p = wrapper.find('p');
      expect( p.text().trim() ).toBe( image.title );
  });

  test('Debe tener una imagen con el alt y el src', () => {
      const img = wrapper.find('img');
      // Mediante prop accedemos a los atributos del elemento html.
      expect( img.prop('src') ).toBe( image.url );
      expect( img.prop('alt') ).toBe( image.title );
  });

  test('Debe tener la clase animate__fadeIn', () => {
      const div = wrapper.find('div');
      // Mediante prop accedemos a los atributos del elemento html.
      // Como hay varias clases, lo sacamos a una variable
      const className = div.prop('className');
      // Mediante includes, buscamos un string dentro de una lista y en caso de que esté, lo evaluamos a true.
      expect( className.includes('animate__fadeIn') ).toBe( true );
  });



})