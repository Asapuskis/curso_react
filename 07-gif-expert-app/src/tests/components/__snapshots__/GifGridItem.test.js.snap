// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <GifGridItem /> Debe mostrar <GifGridItem /> correctamente 1`] = `
<div
  className="card animate__animated animate__fadeIn "
>
  <p>
    hola
  </p>
  <img
    alt="hola"
    src="https://wwhola"
  />
</div>
`;
